src/js/index.js.rst

```diff
@@ -304,19 +379,41 @@ that into the output. Note that the directive lines themselves are not given a
 Blank lines have text consisting of an empty line so they are included in the
 generated JavaScript.
 
+Record the input file line number for each output line so we can create a source
+map later if needed.
+
 .. code-block:: javascript
 
         if (inJavaScript && typeof line.text !== 'undefined') {
           output.push(line.text);
+          inputLineNumbers.push(line.lineNumber);
         }
       });
 
-And ensure the output ends on a newline.
+Create the output, ensuring it ends on a newline.
 
 .. code-block:: javascript
 
-      let result = output.join('\n');
-      result += '\n';
+      output.push('\n');
+      const result = {
+        content: output.join('\n'),
+      };
+
+If the input filename was provided then generate a source map. Without an input
+filename we cannot generate a valid source map. The source map created is line
+based, linking line numbers in the output to corresponding JavaScript block line
+numbers in the input.
+
+.. code-block:: javascript
+
+      if (typeof inputFilename !== 'undefined') {
```

Comment:
![](https://media3.giphy.com/media/4rX9PToX5RrjO/giphy.gif)

Reply:
Yes, Your Holiness, forgive me.

Seriously, this area is a mess and it's been bothering me for a while. The problem is that the original implementation of the command line tool held too close to the Babel tool and provided too many features. Output options to stdout, file, and directory would be fine if the realization of it was tighter.

My thinking is to kill the stdout feature for starters.

For the particular sin at this line, rename the function `fromString` to `fromFileContent` and require the filename parameter. I like the idea of producing sourcemaps by default unless the caller explicitly suppresses them.